File: leetcode/python/198.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        m = nums[0]
        n = max(nums[1], nums[0])

        for i in range(2, len(nums)):
            cur = nums[i]
            if i % 2 == 0:
                m += cur
                m = max(m, n)
            else:
                n += cur
                n = max(m, n)

        return max(m, n)

File: leetcode/java/198.java
class Solution {
  public int rob(int[] nums) {

    int[] dp = new int[nums.length];

    if (nums.length == 2) {
      return Math.max(nums[0], nums[1]);
    } else if (nums.length == 1) {
      return nums[0];
    }

    dp[0] = nums[0];
    dp[1] = Math.max(nums[0], nums[1]);

    for (int i = 2; i < nums.length; i++) {
      dp[i] = Math.max(nums[i] + dp[i - 2], dp[i - 1]);
    }

    int j = nums.length - 1;
    return Math.max(dp[j], dp[j - 1]);
  }
}
